### ISSUE TITLE
[bug] CMakeDeps fails to generate config file for some tool_requires dependencies

### ISSUE THREAD

**Issue by AlexRamallo** [conan-io/conan#12375]
Originally posted here: https://github.com/conan-io/conan-center-index/issues/13704

### Description

I'm not sure if this is a spdlog recipe issue or, or a more general Conan or CMakeDeps/CMakeToolchain issue, but I've only tested this with `spdlog/1.10.0` so far.

If spdlog is present in the dependency graph of the build profile, it will fail to build (see the logs below). This only occurs if spdlog is directly listed in, or is a transient dependency of another package in tool_requires, and **only** if a build and host profile are provided with the `-pr:h/-pr:b` flags.

The issue is reproducible with the `conanio/gcc8` docker image.

I think it might be related to the CMakeToolchain generator because comparing the generated `conan_toolchain.cmake` file from a failed build of spdlog with a working one shows that the failure is due to a missing CMAKE_PREFIX_PATH definition. I.e. a line like the following is missing:

```
set(CMAKE_PREFIX_PATH ${fmt_BUILD_DIRS_RELEASE} ${CMAKE_PREFIX_PATH})
```

### Package and Environment Details

* Package Name/Version: **spdlog/1.10.0**
* Operating System+version: **Linux Ubuntu 18.04.6 LTS**
* Compiler+version: **GCC 8.4.0**
* Docker image: **conanio/gcc8**
* Conan version: **conan 1.53.0**
* Python version: **Python 3.7.13**

### Conan profile

[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=8
compiler.libcxx=libstdc++
build_type=Release
[options]
[conf]
[build_requires]
[env]

### Steps to reproduce

conanfile.txt:
```
[tool_requires]
spdlog/1.10.0
```
install command:
```
conan install . --build=spdlog -pr:b=default -pr:h=default
```

### Logs

<details><summary>Click to expand log</summary>

```
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

spdlog/1.10.0: Forced build from source
conanfile.txt: Installing package
Requirements
Packages
Build requirements
    fmt/8.1.1 from 'conancenter' - Cache
    spdlog/1.10.0 from 'conancenter' - Cache
Build requirements packages
    fmt/8.1.1:1ab440cc9cbac7229437610b68c69ac7438824b1 - Cache
    spdlog/1.10.0:4fc0afed9e517086ec5a7c744ffc2f909aa4aff1 - Build

Installing (downloading, building) binaries...
fmt/8.1.1: Already installed!
spdlog/1.10.0: Configuring sources in /home/conan/.conan/data/spdlog/1.10.0/_/_/source/src
Downloading v1.10.0.tar.gz completed [360.01k]                                           spdlog/1.10.0: 
spdlog/1.10.0: 
spdlog/1.10.0: Copying sources to build folder
spdlog/1.10.0: Building your package in /home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1
spdlog/1.10.0: Generator txt created conanbuildinfo.txt
spdlog/1.10.0: Calling generate()
spdlog/1.10.0: Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
spdlog/1.10.0: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/build/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
spdlog/1.10.0: Aggregating env generators
spdlog/1.10.0: Calling build()
spdlog/1.10.0: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/conan/.conan/data/spdlog/1.10.0/_/_/package/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/src"
-- Using Conan toolchain: /home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/build/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The CXX compiler identification is GNU 8.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build spdlog: 1.10.0
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Build type: Release
CMake Error at CMakeLists.txt:195 (find_package):
  Could not find a package configuration file provided by "fmt" with any of
  the following names:

    fmtConfig.cmake
    fmt-config.cmake

  Add the installation prefix of "fmt" to CMAKE_PREFIX_PATH or set "fmt_DIR"
  to a directory containing one of the above files.  If "fmt" provides a
  separate development package or SDK, be sure it has been installed.


-- Configuring incomplete, errors occurred!
See also "/home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/build/Release/CMakeFiles/CMakeOutput.log".
See also "/home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/build/Release/CMakeFiles/CMakeError.log".
spdlog/1.10.0: 
spdlog/1.10.0: ERROR: Package '4fc0afed9e517086ec5a7c744ffc2f909aa4aff1' build failed
spdlog/1.10.0: WARN: Build folder /home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/build/Release
ERROR: spdlog/1.10.0: Error in build() method, line 116
        cmake.configure()
        ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/conan/.conan/data/spdlog/1.10.0/_/_/package/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/conan/.conan/data/spdlog/1.10.0/_/_/build/4fc0afed9e517086ec5a7c744ffc2f909aa4aff1/src
```

</details>

**Comment by SpaceIm** [comment]
As I said in https://github.com/conan-io/conan-center-index/issues/13704#issuecomment-1288152124, cmake_toolchain content is correct, but CMakeDeps doesn't generate fmt config file.

**Comment by AlexRamallo** [comment]
After some digging it seems like the issue might be related to [build_context_activated](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html#build-context-activated). The transient dependency `fmt` is not being "activated" for some reason, which is why the files don't get generated.

Adding the following line to the spdlog package's generate function fixes the problem for me:
```
cmake_deps.build_context_activated = ["fmt"]
```

The docs are vague about this. Why does this need to be done? Wouldn't you always need config files to be generated for your dependencies?

**Comment by AlexRamallo** [comment]
Also, @SpaceIm let me know if you'd like me to send a PR with that change for the spdlog package, or if this should still be something resolved in the conan client.

**Comment by SpaceIm** [comment]
`spdlog` recipe shouldn't be "fixed". IMHO, it's a bug in conan client.
consumers shouldn't have to enable `build_context_activated` to fix this issue. The purpose of `build_context_activated` is to generate config files of `tool_requires` of a recipe. Here `fmt` is a `requires` of `spdlog`, so its config file should be generated in the context of `spdlog` build, regardless of `build_context_activated` value. The fact that the build of spdlog has been triggered indirectly through the fact that it's a `tool_requires` or `requires` of a conanfile should not matter.

**Comment by ericLemanissier** [comment]
This should have been fixed with https://github.com/conan-io/conan/pull/12675 which will be included in conan 1.56
